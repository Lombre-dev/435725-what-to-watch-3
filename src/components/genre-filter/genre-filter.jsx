import PropTypes from 'prop-types';
import React from 'react';
import {connect} from 'react-redux';
import {ActionCreator} from '../../reducer';

class GenreFilter extends React.PureComponent {

  constructor(props) {
    super(props);

    this._handleItemClick = this._handleItemClick.bind(this);
  }

  _handleItemClick(e) {

    const {onSelect} = this.props;
    const genre = e.currentTarget.dataset.genre;

    e.preventDefault();
    onSelect(genre);
  }

  render() {

    const {genres, currentGenre} = this.props;

    return (
      <ul className="catalog__genres-list">
        {
          genres.map((tab) => {
            return (
              <li key={tab}
                className={`catalog__genres-item${tab === currentGenre ? ` catalog__genres-item--active` : ``}`}
              >
                <a
                  data-genre={tab}
                  href="#"
                  className="catalog__genres-link"
                  onClick={this._handleItemClick}
                >{tab}</a>
              </li>
            );
          })
        }
      </ul>
    );
  }
}

GenreFilter.propTypes = {
  genres: PropTypes.arrayOf(PropTypes.string).isRequired,
  currentGenre: PropTypes.string.isRequired,
  onSelect: PropTypes.func.isRequired,
};

/*
TODO: решил проваливать эти пропсы сверху,
если лучше привязаться к store, то пометь в feedback.
function mapStateToProps(state) {
  return {
    genres: state.catalogGenres,
    currentGenre: state.cataloGenre,
  };
}
*/

function mapDispatchToProps(dispatch) {
  return {
    onSelect: (genre) => {
      dispatch(ActionCreator.setCatalogGenre(genre));
    }
  };
}

export {GenreFilter};
export default connect(null, mapDispatchToProps)(GenreFilter);
