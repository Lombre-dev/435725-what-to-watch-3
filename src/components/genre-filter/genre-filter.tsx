import * as React from 'react';
import {connect} from 'react-redux';
import {setCatalogGenre} from '../../redux/catalog/actions';
import {getGenreLabels} from '../../utils/movie-utils';

type TGenreFilterProps = {
  genre: string;
  onSelect: Function;
};

class GenreFilter extends React.PureComponent<TGenreFilterProps> {

  public constructor(props: TGenreFilterProps) {
    super(props);

    this._handleClick = this._handleClick.bind(this);
  }

  private _handleClick(e: React.SyntheticEvent) {

    const {genre, onSelect} = this.props;

    e.preventDefault();
    onSelect(genre);
  }

  public render() {

    const {genre} = this.props;
    const genreLabels = getGenreLabels();

    return (
      <a
        href="#"
        className="catalog__genres-link"
        onClick={this._handleClick}
      >{genreLabels[genre] || genre}</a>
    );
  }
}

function mapDispatchToProps(dispatch: Function) {
  return {
    onSelect: (genre: string) => {
      dispatch(setCatalogGenre(genre));
    },
  };
}

export {GenreFilter};
export default connect(null, mapDispatchToProps)(GenreFilter);
