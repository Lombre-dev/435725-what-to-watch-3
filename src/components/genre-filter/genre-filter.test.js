import React from 'react';
import renderer from 'react-test-renderer';
import {GenreFilter} from './genre-filter';

const GENRES = [`Comedy`, `Sci-Fi`, `Horror`];
const HANDLE_SELECT = () => {};

describe(`<GenreFilter />`, () => {

  it(`render should be match markup`, () => {

    const result = renderer
      .create(<GenreFilter
        genres={GENRES}
        currentGenre={GENRES[0]}
        onSelect={HANDLE_SELECT}
      />)
      .toJSON();

    expect(result).toMatchSnapshot();
  });
});
