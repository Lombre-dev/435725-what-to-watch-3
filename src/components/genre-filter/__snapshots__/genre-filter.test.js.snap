// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GenreFilter /> render should be match markup 1`] = `
<ul
  className="catalog__genres-list"
>
  <li
    className="catalog__genres-item catalog__genres-item--active"
  >
    <a
      className="catalog__genres-link"
      data-genre="Comedy"
      href="#"
      onClick={[Function]}
    >
      Comedy
    </a>
  </li>
  <li
    className="catalog__genres-item"
  >
    <a
      className="catalog__genres-link"
      data-genre="Sci-Fi"
      href="#"
      onClick={[Function]}
    >
      Sci-Fi
    </a>
  </li>
  <li
    className="catalog__genres-item"
  >
    <a
      className="catalog__genres-link"
      data-genre="Horror"
      href="#"
      onClick={[Function]}
    >
      Horror
    </a>
  </li>
</ul>
`;
